docs/design/topics.md

```diff
@@ -0,0 +1,168 @@
+also replace the current bookmark concept for Git interop. As they have been
+documented to be confusing users coming from Git. They also supersede the 
+`[experimental-advance-branches]` config for those who currently use it, as 
+such a behavior will be built-in for Topics.
+
+Topics have been discussed heavily since their appearance in 
+[this Discussion][gh-discuss]. As Noah, Anton and I had a long 
+[Discord discussion][dc-thread] about them, which then also poured into the 
+[Topic issue][issue].
+
+## Prior work 
+
+Currently there only is Mercurial which has a implementation of 
+[Topics][hg-topics]. There also is the [Topic feature][gerrit-topics] in Gerrit,
+which groups commits with a single identifier.
+
+
+## Goals and non-goals
+
+### Goals
+
+The goals for this Project are small, see below.
+
+* Introduce the concept of native topological branches for Jujutsu.
+* Simplify Git interop by reducing the burden on Bookmarks.
+* Add Change metadata as a storage concept.
+* Remove the awkward `bookmark` to Git `branch` mapping.
+
+### Non-Goals
+
+* Making bookmarks unnecessary.
+
+## Overview
+
+Until now, Jujutsu had no native set of topological branches, just 
+[Bookmarks][bm] which interact poorly with Git's expectation of branches. 
+Topics on the otherhand are can be made to represent Git branches as users 
+expect them, see [Julia Evans poll][jvns-poll]. They also allow us to 
+seamlessly take over the [tracking-branches][tb] concept.
+
+Other use-cases they're useful for are representing a set of 
+[archived commits][archived] or even a [checkout history][checkout].
+
+
+#### Behavior
+
+A Topic is a set of Changes marked with a name, which infectiously moves 
+forward with each descendant you create. All Changes without a named topic are
+implicitly in a anonymous topic, which gets separately tracked as soon as you 
+send it in for review (so materialize it as a Git branch). A Topic may have
+multiple heads. A single Change may belong to multiple Topics, which means 
+that they also can get exported as overlapping Git branches. 
+
+
+TODO: Example here
+
+#### Major Use-Cases 
+
+
+##### Obsoletion of `[experimental.auto-advance-bookmarks]`
+
+Since Topics are infectious by nature, they can perfectly map to a Git branch.
+This alleviates the need for the auto-advancing bookmarks. It also makes the 
+use-case of the solo developer working on Git `main` easier, as you just mark
+all descendants of `trunk()` belonging to the `main` topic, which then gets 
+translated to the `main` branch on `jj git push`.
+
+
+##### Change Archival
+
+For the archival use-case the infectious property of a Topic isn't as 
+important. So having a 
+`jj topic create -r 'trunk()..@ & description("archive:")' should mark them all
+revisions with a `archive:` description as archived. For this use-case the 
+non-contiguous property of Topics really shine.
+
+##### Checkout history
+
+### Detailed Design
+
+#### Git Interop
```

Comment:
There's a part of this that seems to be missing: where does a topic "end"? I'm guessing that you would export one git branch for each head commit in the topic, rather than one git branch for _every commit_ in the topic (since that wouldn't really have much advantage compared to `jj git push -c`). But git doesn't have a concept of where a branch "starts" except relative to other branches, and then only in the context of commands (such as rebase) that start by finding common ancestors, which implicitly "prune" the branch. So for the "change-archival" use-case, what commits would be included in the `archive` topic?

I'm guessing the answer is something like: `topic(x) := ::heads(topic) ~ ::trunk()`, i.e. topic `x` contains the heads plus all ancestors _minus_ everything on the `trunk` "branch". I think this is consistent with the property of `hg` topics that they "disappear" as soon as they are merged (though I haven't used `hg` and am not totally sure I followed the linked `topic` documentation). But whatever the rule is, it should be explicit.

Reply:
> There's a part of this that seems to be missing: where does a topic "end"? I'm guessing that you would export one git branch for each head commit in the topic, rather than one git branch for _every commit_ in the topic (since that wouldn't really have much advantage compared to `jj git push -c`).

Yes, this is also correct. 

> So for the "change-archival" use-case, what commits would be included in the `archive` topic?

This is @avamsi's use-case from #4180. This means that you locally add a description like `archive:` to your commits and then let `jj` infer it in someway or occasionally do it manually with a command. 

> I'm guessing the answer is something like: `topic(x) := ::heads(topic) ~ ::trunk()`, i.e. topic `x` contains the heads plus all ancestors _minus_ everything on the `trunk` "branch".

Yes, that is correct and thanks for the formulation. 

> But whatever the rule is, it should be explicit.

Yes, I'll write it down.